rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,566 @@
+---
+feature: feature_parameter_names
+start-date: 2024-01-10
+author: Dmitry Bogatov
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs has mechanisms to exports compile-time options of the upstream build
+system through the named parameters of derivation function, but the names of
+these parameters are inconsistent from package to package and from feature to
+feature.
+
+# Motivation
+[motivation]: #motivation
+
+Developer interface for building environment with specific features enabled and
+disabled is more complicated and and involves more "you just have to know it"
+compared to already existing systems.
+
+1. Knowing that package has a feature parameter (out of knowning the upstream)
+   does not mechanically translates to the name of feature parameter, it can be
+   named in multiple ways.
+
+2. Writing generic code is complicated. Even if list of packages all have
+   optional dependency on package `foo`, duck-typing like
+
+```
+map (p: p.override { with_foo = false; }) packages
+```
+
+won't work because of lack of uniformity, leading ugly code downstream.
+
+3. If user wants to globally disable support for optional feature or build
+   dependency as much as possible, he needs to set multiple feature parameters,
+   and he still can miss couple. E.g, it is not obvious whether the following
+   list is enough to cover everything X11-related:
+
+```
+withX
+withX11
+x11Support
+enableX
+```
+
+4. There is no clear separation between feature parameters and other arguments
+   to the package. With naming mandated by this RFC, distinction is clear to
+   cater to automatic queries.
+
+5. As side effect of the proposad migration plan is compiling exahustive list
+   of known feature parameters. It is not huge stretch to have CI check to
+   enforce that this list is current, bringing nixpkgs on par with Gentoo
+   [use-flags](https://www.gentoo.org/support/use-flags).
```

Comment:
Doesn't Gentoo separately invest effort in consistently providing the popular USE flags for almost all the  packages where they make any amount of sense?

Reply:
At some point did some manual comparison between Gentoo and Nixpkgs and learned that Nixpkgs exports no less of feature parameters than Gentoo, just in less uniform way. In other words, I expected Gentoo configuration to be more granular than Nixpkgs, but that wasn't the case.

I'll move the last sentence into the future work to not inadvertently imply that RFC demands Nixpkgs to have as much feature parameters as Gentoo has use flags.

Reply:
My expectation was that Nixpkgs has a comparable number of feature parameters, but the corresponding feature parameters are provided by fewer packages, isn't it so?

Reply:
Probably. Many packages do something like

```
buildInputs = lib.optionals stdenv.isLinux [ foo ];
```

without bothering to introduce `with_foo` flag.